Streamer input source prefetching (12_5_X)

smorovic opened the pull request:
#### PR description:

Adds a preallocated buffer intended to avoid many small reads when reading streamer files. This is aimed at improving EOS reading performance (see https://github.com/cms-sw/cmssw/issues/38997).
A pointer to the big buffer is passed if possible for reading header and even payloads (contiguous), otherwise data is copied to  the intermediate header or event buffer.

New parameter enables and configures size of the large buffer in MBytes (0 to disable), for example:
```
process.source = cms.Source("NewEventStreamFileReader",
    prefetchMBytes = cms.untracked.uint32(100),
```

#### PR validation:

A new unit test in `IOPool/Streamer` specifically checks behavior with this parameter (1 MB with small generated events). A file larger than the buffer was created to test behavior at the buffer boundary.
Internally, source checks adler32 checksum of both the header and event payload (compared to a header field), so corruption of payloads would be caught by unit tests.


Streamer source in `CalibCalorimetry/EcalLaserSorting` uses this code and has a respective unit test in that location.
Similarly, DQM streamer input source is tested as part of DAQ I/O chain in `EventFilter/Utilities` unit test.

smorovic committed:
prefetch configurable buffer size per posix read in the Streamer input source

smorovic committed:
use original buffer when possible

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31600

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31600/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31600/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

smorovic committed:
code-format

smorovic commented:
12_4_X backport branch (PR will be created after this is reviewed):
https://github.com/smorovic/cmssw/tree/12_4_X-streamer-prebuffering

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31601

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @smorovic (Srecko Morovic) for master.

It involves the following packages:

- IOPool/Streamer (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@makortel, @wddgit this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

smorovic commented:
@cmsbuild please test

smorovic committed:
add unit test file

smorovic commented:
abort test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31604

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #39088 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

smorovic commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b93f17/26872/summary.html
**COMMIT**: 99709c7e462cbb0496aaf1e7848846345fa52d63
**CMSSW**: CMSSW_12_5_X_2022-08-16-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39088/26872/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 136.9 step 4
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3692500
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3692464
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-16-2300+b93f17/52178/triggerResults): no differences found

Dr15Jones commented on IOPool/Streamer/interface/StreamerInputFile.h:
```diff
@@ -51,7 +54,7 @@ namespace edm {
   private:
     void openStreamerFile(std::string const& name, std::string const& LFN);
-    storage::IOSize readBytes(char* buf, storage::IOSize nBytes);
+    storage::IOSize readBytes(char** buf, storage::IOSize nBytes, bool zeroCopy = true, unsigned skipped = 0);
```
I think using `char*& buf` would be cleaner as it better expresses what is being done.

Dr15Jones replied:
Looking further at the code, I actually think the better solution would be to change the signature to return the starting offset, 
```
std::pair<storage::IOSize, char*> readBytes(char* buf, storage::IOSize nBytes, bool zeroCopy = true, unsigned skipped = 0)
```

That way you don't have to worry about accidentally modifying the address of the buffer which seems to be why you were forced not no longer call using the argument `&eventBuf_[sizeof(EventHeader)]`

smorovic replied:
What complicated it here it is that the header is read first, so the offset moves by that amount, then creation of the message view in the end needs everything from the start of the header.
`&eventBuf_[sizeof(EventHeader)]` is a single pointer (and not a variable), so it can't work here. Instead I pass **buf which points to the beginning of the small buffer. If there is no zero-copy, I copy payload to the buffer at offset skippedHdr. If there is zero-copy, I overwrite the pointer value to the position of big buffer offset - skippedHdr (so it starts at the header).

As you suggest, we could areturn the final pointer in a pair and then call xxPtr = pair.second, instead of passing `char **` or `char *&` to the function.

I just realized there may be another problem - resize can reallocate vector buffer in heap, so `headerPtr = &headerBuf_[0];` and `eventPtr = &eventBuf_[0];` should be called again after resize (of the small buffer).

Dr15Jones commented on IOPool/Streamer/src/StreamerInputFile.cc:
```diff
@@ -257,7 +298,7 @@ namespace edm {
       eventRead = true;
       if (eventSkipperByID_) {
-        EventHeader* evh = (EventHeader*)(&eventBuf_[0]);
+        EventHeader* evh = (EventHeader*)(eventPtr);
```
Is this guaranteed to have the correct byte alignment now?

smorovic replied:
Here eventPtr isn't changed (using zeroCopy=false for reading the event header in previous step) and still points to the beginning of the std::vector allocation, so it should be ok.

smorovic replied:
Now I see that there is the same problem with the EventMsgView:
```
uint32 EventMsgView::protocolVersion() const {
  EventHeader* h = (EventHeader*)buf_;
  return h->protocolVersion_;
}
```
so it's not safe to use non-aligned starting address and we need to copy the event....
It's still safe for the InitMessage, because it's always at the beginning of the buffer.
For event it could be enabled if event size padding is introduced to the streamer format.

Dr15Jones commented on IOPool/Streamer/src/StreamerInputFile.cc:
```diff
@@ -111,10 +119,40 @@ namespace edm {
+      if (prefetchMBytes_) {
+        //assert(tempPos_ > tempLen_);
+        if (tempPos_ == tempLen_) {
+          n = storage_->read(&tempBuf_[0], prefetchMBytes_ * 1024 * 1024);
```
maybe 
```suggestion
          n = storage_->read(&tempBuf_[0], tempBuf_.size());
```

Dr15Jones commented on IOPool/Streamer/src/StreamerInputFile.cc:
```diff
@@ -171,7 +210,7 @@ namespace edm {
     }
 
-    startMsg_ = std::make_shared<InitMsgView>(&headerBuf_[0]);  // propagate_const<T> has no reset() function
+    startMsg_ = std::make_shared<InitMsgView>(headerPtr);  // propagate_const<T> has no reset() function
```
Isn't there a problem here (and maybe there was a problem with this code before this change as well)? `startMsg_` stays around until the next file is read. But it is internally holding a pointer to what `headerPtr` was addressing at this point. However, with this change you are using `noCopy` so `headerPtr` in this routine points to something in `tmpBuffer_` which will change so when `startMsg_` is read at next file boundary, the memory it reads will not be the same as when it was called here.

smorovic replied:
In StreamerInputFile it's only used to get currRun_ and currProto_, they are fetched when the message is new (after `readStartMessage`), so that is ok. There is also a check if smart pointer of InitMsgView isn't null in `openNextFile`, but that doesn't rely on the tmpBuffer_.
And in `StreamerFileReader.cc` the header is always accessed and deserialized after opening a new fileand not kept around for later.
So I think it's ok, but I will have another look.

smorovic replied:
After anothe look, it is safe. Also, a unit test job now opens two files larger than the prefetch buffer successively, so a corrupted INI would be detected.

smorovic committed:
Update IOPool/Streamer/src/StreamerInputFile.cc

Using the actual buffer size

Co-authored-by: Chris Jones <chrisdjones15@gmail.com>

smorovic committed:
* fix unsafe usage of a pointer after potential resize of the vector,
which can relocate the buffer in heap

* expand unit test and process twice the file larger than the prefetch
 buffer (to check if there is any problem with zero-copy handling of the
 header)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31615

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #39088 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31616

- This PR adds an extra 24KB to repository

makortel commented on IOPool/Streamer/src/StreamerInputFile.cc:
```diff
@@ -140,13 +178,14 @@ namespace edm {
     }
 
-    HeaderView head(&headerBuf_[0]);
+    HeaderView head(headerPtr);
```
Could I suggest to enclose this and the following lines into a code block so that it is clear that `head` can not be used after the possible  `headerBuf_.resize()`? I.e.
```cpp
    {
      HeaderView head(headerPtr);
      uint32 code = head.code();
      if (code != Header::INIT) /** Not an init message should return ******/
      {
        throw Exception(errors::FileReadError, "StreamerInputFile::readStartMessage")
            << "Expecting an init Message at start of file\n";
        return;
      }
    }
```

(same comment for the similar pattern in `readEventMessage()`)

makortel commented on IOPool/Streamer/test/RunSimple_NewStreamer.sh:
```diff
@@ -31,12 +31,14 @@ cd ${OUTDIR}
 cmsRun NewStreamOut_cfg.py compAlgo=${TEST_COMPRESSION_ALGO} > out 2>&1 || die "cmsRun NewStreamOut_cfg.py compAlgo=${TEST_COMPRESSION_ALGO}" $?
 cmsRun NewStreamOutAlt_cfg.py compAlgo=${TEST_COMPRESSION_ALGO} > outAlt 2>&1 || die "cmsRun NewStreamOut_cfg.py compAlgo=${TEST_COMPRESSION_ALGO}" $?
 cmsRun NewStreamOutExt_cfg.py compAlgo=${TEST_COMPRESSION_ALGO} > outExt 2>&1 || die "cmsRun NewStreamOut_cfg.py compAlgo=${TEST_COMPRESSION_ALGO}" $?
+cmsRun NewStreamOutExt2_cfg.py compAlgo=${TEST_COMPRESSION_ALGO} > outExt 2>&1 || die "cmsRun NewStreamOut_cfg.py compAlgo=${TEST_COMPRESSION_ALGO}" $?
```
```suggestion
cmsRun NewStreamOutExt2_cfg.py compAlgo=${TEST_COMPRESSION_ALGO} > outExt 2>&1 || die "cmsRun NewStreamOutExt2_cfg.py compAlgo=${TEST_COMPRESSION_ALGO}" $?
```
(if you feel like it, `Alt` and `Ext` could be added to the two lines above)

How about also adding a check that the `CHECKSUM`  printout is as expected? (I see also this check is missing for the `Alt` and `Ext` tests)

smorovic replied:
Alt and Ext combine multiple input files, so the check can not be done (it is comparing products between a single input or output process).
Anyway, streamer reader checks the header and each event for adler32 checksum consistency.

smorovic committed:
*readBytes returning data buffer pointer
*disable zero-copy for event buffer as it is not aligned address
* enclose header access before buffer vector is resized

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31617

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31617/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31617/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31617/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31617/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

smorovic committed:
Update IOPool/Streamer/test/RunSimple_NewStreamer.sh

Co-authored-by: Matti Kortelainen <matti.kortelainen@cern.ch>

smorovic committed:
code format fix

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31618

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31618/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31618/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31618/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31618/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31620

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31620/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31620/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

smorovic committed:
fix code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31628

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #39088 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

smorovic commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b93f17/26891/summary.html
**COMMIT**: 3db4eedaf79dafc7d94272ac61199ee4c329044f
**CMSSW**: CMSSW_12_5_X_2022-08-17-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39088/26891/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 136.9 step 4
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3692500
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3692464
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-17-1100+b93f17/52192/triggerResults): no differences found

makortel commented on IOPool/Streamer/src/StreamerInputFile.cc:
```diff
@@ -40,21 +43,26 @@ namespace edm {
         currentFileName_(),
         currentFileOpen_(false),
         eventSkipperByID_(eventSkipperByID),
+        prefetchMBytes_(prefetchMBytes),
```
It seems to me the `prefetchMbytes_` has now become mostly unnecessary. The check
https://github.com/cms-sw/cmssw/blob/3db4eedaf79dafc7d94272ac61199ee4c329044f/IOPool/Streamer/src/StreamerInputFile.cc#L129
could be replaced with
```cpp
if (not tempBuf_.empty()) {
```

smorovic replied:
Good point. I removed prefetchMBytes_.

makortel commented on IOPool/Streamer/src/StreamerInputFile.cc:
```diff
@@ -111,16 +119,48 @@ namespace edm {
+                                                                 bool zeroCopy,
+                                                                 unsigned int skippedHdr) {
     storage::IOSize n = 0;
+    char* ptr = buf - skippedHdr;
```
After spending some time in wondering why `ptr` is initialized to `buf - skippedHdr` I'd suggest to add a comment that in case of successful read (`n == nBytes`), the caller expects the returned pointer to point to the beginning of the header.

smorovic replied:
Added:
```
    //returned pointer should point to the beginning of the header
    //even if we read event payload that comes afterwards
```
and a couple of more comments in that function.

smorovic committed:
* improve comments for the readBytes function
* remove unnecessary class member
* allow zero-copy event reading because EventHeader and EventMessage consist of byte and byte array variables and non-aligned casting is safe

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31666

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31666/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31666/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31666/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31666/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31667

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31667/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31667/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

smorovic committed:
apply code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39088/31669

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #39088 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

smorovic commented:
Apparently the Event Header
https://github.com/cms-sw/cmssw/blob/master/IOPool/Streamer/interface/EventMessage.h#L61
and also this header which is part of it:
https://github.com/cms-sw/cmssw/blob/master/IOPool/Streamer/interface/MsgHeader.h#L6
consist fully of of char and char arrays which are defined here:
https://github.com/cms-sw/cmssw/blob/master/IOPool/Streamer/interface/MsgTools.h#L15

EventMsgView uses char* pointer to traverse through the message:
https://github.com/cms-sw/cmssw/blob/master/IOPool/Streamer/src/EventMessage.cc#L4
and builds higher-order variables from individual bytes (like with the convert32 function).

Conclusion is that alignment is not an issue. So, in the latest commit I enabled 'zero-copy' mode parameter for reading events.
It works fine both in the unit test and reading a production streamer file larger than 100 MB.

smorovic commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b93f17/26967/summary.html
**COMMIT**: d6ea8aab85309c63f5c030442fb4cc281600eab9
**CMSSW**: CMSSW_12_5_X_2022-08-21-0000/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39088/26967/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3693040
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3693016
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-21-0000+b93f17/52251/triggerResults): no differences found

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

makortel commented:
urgent

Just for @cms-sw/orp-l2 to note that it would be good to have this PR in pre5 so that Tier0 could potentially test the functionality soon (before it becomes part of a production release).

perrotta commented:
+1
